### PR TITLE
New lint for struct update that has no effect

### PR DESCRIPTION
**Pull request by sanxiyn**
This is my first lint to Clippy. It is an implementaiton of rust-lang/rust#23199.

It does not implement side-effect check yet (see the Rust issue), but I think maybe Clippy should warn when the update has no effect except side-effect?

### CODE REVIEW

**Comment by llogiq** [comment]
Please run `util/update_lints.py` and commit README.md to make the build pass.

File: tests/compile-fail/needless_update.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+struct S {
+    pub a: i32,
+}
+
+fn main() {
+    let base = S { a: 0 };
+    S { ..base }; // no error
+    S { a: 1, ..base }; //~ERROR update has no effect
+}
```
Comment (llogiq) [inline tests/compile-fail/needless_update.rs]: Nit: I'd like to see a test case with more than one field and 0, 1, all fields updated.

**Comment by Manishearth** [comment]
I like this lint idea :smile: 

Not sure if we need the side effect check. You can always hoist the expression out of the struct.

Otherwise, linting only when the expression is a simple ExprPath works.

**Comment by sanxiyn** [comment]
Updated README.md and the test.

**Comment by sanxiyn** [comment]
Updated the message.

**Comment by Manishearth** [comment]
Thanks!

**Pull request merged**